### DISCUSSION TITLE
Is it possible to pass variables to private functions?

### DISCUSSION THREAD

**Discussion by tiymat** [lmorg/murex#695] [Q&A] [answered]
Hi all! I'm wondering if anyone knows if it's possible to pass data (in my case, a string) to a private function.

In the documentation for `private` it looks like it doesn't support parameterizing variable names like `function` does, and indeed when I try doing so I get an error saying the private is not a valid code block.

I've also tried writing the value into a named pipe, invoking the private, and then reading from that pipe from within the private, but I always get empty output back. But I'm still not completely comfortable with named pipes, so I could be doing something wrong there. Any advice would be welcomed!

Thanks,
-tiymat

**Comment by lmorg** [comment, marked as answer]
You can pass parameters the same as any command, eg
```
private pvtExample {
    $PARAMS
    out "second param is $2"
}

function example {
    pvtExample foo bar
}
```
running example there will return
```
[
    "foo",
    "bar"
]second param is bar
```
At the moment these are best documented on our dev website:
- https://dev.murex.rocks/variables/PARAMS.html
- https://dev.murex.rocks/variables/numeric.html
(as an aside, it would be interesting to hear your feedback on this updated website)

There is a pending ticket to bring named variables to privates too:
https://github.com/lmorg/murex/issues/429

---

Regarding named pipes, yeah they're fiddly as hell. I'm going to revisit them at some point soon and see what I can do to improve them.

**Comment by tiymat** [comment]
Thanks so much! Got it working, after reading your answer -- my mistake was piping `$a -> foo` and expecting `$1` in foo to equal `$a` within `foo`, rather than invoking it like `foo $a`.

---

Re: the website, it looks very promising, albeit with a couple of kinks to iron out.

I really like the menubar on the left with the dropdowns that allow quick navigation of the site. Very often my workflow on the old version would be to read up on a command or language feature, learn something that would make me want to know something about a different topic not linked in the "See Also" section, open a new browser tab to search the topic, locate it in the search results, and then open the page.

This isn't the worst thing in the world, but it's a bit clunky. In my limited testing on the new site I've found I'm consistently able to locate most things I want to know about in the left menubar quickly, and middle-click the link to open it in a new tab, which I'm finding way more convenient.

Also, maybe this is just personal preference, but just having a list of all the data types, or all the builtins, available at a glance makes me feel much more like I have a grip on the features available to me and how I might use them together. (e.g. I noticed the recent(?) addition of the signalTrap event in the menu, when on the old site it might have taken me a long time to happen to stumble across the events page.)

---

There are some changes I'd make. (It's possible some of these issues might be the site breaking due to my relatively uncommon browser choice (Pale Moon) and not issues that will affect 99% of users.)

At the [PARAMS](https://dev.murex.rocks/variables/PARAMS.html) page you linked, I clicked the "Variables" button above the page title and was brought to [this page](https://dev.murex.rocks/variables/) which doesn't have an HTML title element defined and whose "Go back" and "Take me home" buttons do nothing. The links to the different pages about variables do work though. I was able to reproduce this same kind of pattern with a few different pages, like [this one](https://dev.murex.rocks/types/) which also has a misaligned "On this page" section.

I'm not sure if it's intentional that the tags below the page title and author aren't clickable, or just a bug, but c'mon, you can't put all those bright, button-like elements at the top of the page and expect me not to try to click them! :grin:
Though given how convenient navigation is through the menu, it wasn't a big deal.

Also, the most minor of minor nitpicks, I'd suggest changing `Documentations` in the menu bar running along the top to `Documentation`.

**Reply by lmorg** [reply]
great feedback thank you.
